plugins/yaml/src/org/jetbrains/yaml/references/YAMLAliasReference.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2000-2017 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jetbrains.yaml.references;
+
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiReferenceBase;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.intellij.util.ArrayUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.yaml.psi.YAMLScalar;
+
+import java.util.Collection;
+import java.util.Comparator;
+
+public class YAMLAliasReference extends PsiReferenceBase<PsiElement> {
+
+  private static final String ANCHOR_PREFIX = "&";
+
+  public YAMLAliasReference(@NotNull PsiElement element) {
+    super(element);
+  }
+
+  @Nullable
+  @Override
+  public PsiElement resolve() {
+    Collection<YAMLScalar> values = PsiTreeUtil.findChildrenOfType(myElement.getContainingFile(), YAMLScalar.class);
```

Comment:
This approach works correctly, however, from my point of view, it solves only the resolution task properly while some further features based on it will work slow, basically O(n^2) in some cases where n ~ file size.

I think what we need to add here is proper lexing of anchors and refs (aliases) like tags are being lexed right now (see `org/jetbrains/yaml/lexer/yaml.flex:161` and usages); it will eliminate the need to instantiate text of the scalars in the whole document.

What do you think?
If you don't like the idea of digging into lexer and parser, I think I can help with that :)